### PR TITLE
update s3 ListObjects to V2

### PR DESCRIPTION
**Pull request by davidwtf**
Signed-off-by: Tengfei Wang <tfwang@alauda.io>

When using [minio](https://min.io/) as s3 provider, `ListObjects` may meet 502 error on certain conditions, but `ListObjectsV2` works well.

### CODE REVIEW

**Comment by techknowlogick** [comment]
Please note, this will cause those hosting w/ DO Spaces to fail, as DO doesn't support `ListObjectsV2`

File: registry/storage/driver/s3-aws/s3.go [hunk 1]
```diff
@@ -667,13 +667,13 @@ func (d *driver) List(ctx context.Context, opath string) ([]string, error) {
...
 		}
 
-		if *resp.IsTruncated {
-			resp, err = d.S3.ListObjects(&s3.ListObjectsInput{
-				Bucket:    aws.String(d.Bucket),
-				Prefix:    aws.String(d.s3Path(path)),
-				Delimiter: aws.String("/"),
-				MaxKeys:   aws.Int64(listMax),
-				Marker:    resp.NextMarker,
+		if resp.IsTruncated != nil && *resp.IsTruncated {
```
Comment (arkodg) [inline registry/storage/driver/s3-aws/s3.go:670]: `&& *resp.IsTruncated` is not required

**Comment by arkodg** [comment]
@davidwtf can you please add a test to 
https://github.com/docker/distribution/blob/master/registry/storage/driver/s3-aws/s3.go

**Comment by milosgajdos** [comment]
Hey @davidswu, there is a good chance might get merged soon. Would be awesome if you could find some spare time to address the points @arkodg raised.

File: registry/storage/driver/s3-aws/s3.go (hunk 1 above)
Comment (davidwtf) [reply registry/storage/driver/s3-aws/s3.go:670]: I think it's better to keep it, considering that S3 compatible servers may return false.

**Comment by davidwtf** [comment]
> Hey @davidwtf , there is a good chance might get merged soon. Would be awesome if you could find some spare time to address the points @arkodg raised.

Sorry for my late reply, I just added a test function `TestListObjectsV2` and pushed again.

@arkodg

**Review by arkodg — APPROVED** [review]
LGTM, thanks !

File: registry/storage/driver/s3-aws/s3.go [hunk 2]
```diff
@@ -667,13 +667,13 @@ func (d *driver) List(ctx context.Context, opath string) ([]string, error) {
...
+			resp, err = d.S3.ListObjectsV2(&s3.ListObjectsV2Input{
+				Bucket:            aws.String(d.Bucket),
+				Prefix:            aws.String(d.s3Path(path)),
+				Delimiter:         aws.String("/"),
+				MaxKeys:           aws.Int64(listMax),
+				ContinuationToken: resp.ContinuationToken,
```
Comment (milosgajdos) [inline registry/storage/driver/s3-aws/s3.go:676]: This must be set to `resp.NextContinuationToken`

As per [AWS docs](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html):

> NextContinuationToken
NextContinuationToken is sent when isTruncated is true, which means there are more keys in the bucket that can be listed. The next list requests to Amazon S3 can be continued with this NextContinuationToken. NextContinuationToken is obfuscated and is not a real key

File: registry/storage/driver/s3-aws/s3.go [hunk 3]
```diff
@@ -834,8 +834,7 @@ ListLoop:
 			})
 		}
 
-		// resp.Contents must have at least one element or we would have returned not found
-		listObjectsInput.Marker = resp.Contents[len(resp.Contents)-1].Key
+		listObjectsInput.ContinuationToken = resp.ContinuationToken
```
Comment (milosgajdos) [inline registry/storage/driver/s3-aws/s3.go:837]: What is `resp.ContinuationToken` set to when the initial request does not include it?

Per [AWS Docs](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html):

> ContinuationToken
If ContinuationToken was sent with the request, it is included in the response.

**Review by milosgajdos — COMMENTED** [review]
PTAL @davidwtf

File: registry/storage/driver/s3-aws/s3.go (hunk 2 above)
Comment (davidwtf) [reply registry/storage/driver/s3-aws/s3.go:676]: yes, you are right, I' ll fix it.

File: registry/storage/driver/s3-aws/s3.go (hunk 3 above)
Comment (davidwtf) [reply registry/storage/driver/s3-aws/s3.go:837]: same problem, I'll fix it.

File: registry/storage/driver/s3-aws/s3.go (hunk 2 above)
Comment (davidwtf) [reply registry/storage/driver/s3-aws/s3.go]: commit again.
PTAL

**Comment by milosgajdos** [comment]
I think you need to rebase now @davidwtf

**Comment by davidwtf** [comment]
> I think you need to rebase now @davidwtf 

rebased

**Comment by sargun** [comment]
Any reason you don't use the pagination helpers as used in the walk API?

**Comment by davidwtf** [comment]
> Any reason you don't use the pagination helpers as used in the walk API?

This pr just change ListObjects to ListObjectsV2. I don't know why the previous code doesn't use pagination.

File: registry/storage/driver/s3-aws/s3_test.go [hunk 4]
```diff
@@ -357,3 +360,88 @@ func TestMoveWithMultipartCopy(t *testing.T) {
...
+		t.Fatalf("unexpected error listing: %v", err)
+	}
+
+	sort.Strings(subPathes)
+	sort.Strings(result)
+	if !reflect.DeepEqual(subPathes, result) {
+		t.Fatalf("unexpected list result")
+	}
+
+	var walkPathes []string
```
Comment (milosgajdos) [inline registry/storage/driver/s3-aws/s3_test.go:420]: Can you please rename this variable to `walkPaths`

File: registry/storage/driver/s3-aws/s3_test.go [hunk 5]
```diff
@@ -357,3 +360,88 @@ func TestMoveWithMultipartCopy(t *testing.T) {
...
+			if fileInfo.IsDir() || fileInfo.Size() != int64(len(fileInfo.Path())) {
+				t.Fatalf("unexpected walking file info")
+			}
+		}
+		return nil
+	}); err != nil {
+		t.Fatalf("unexpected error walking: %v", err)
+	}
+
+	subPathes = append(subPathes, subDirPath)
```
Comment (milosgajdos) [inline registry/storage/driver/s3-aws/s3_test.go:437]: Can you please rename this variable to `subPaths`?

File: registry/storage/driver/s3-aws/s3_test.go [hunk 6]
```diff
@@ -357,3 +360,88 @@ func TestMoveWithMultipartCopy(t *testing.T) {
...
+
+	d, err := s3DriverConstructor(rootDir, s3.StorageClassStandard)
+	if err != nil {
+		t.Fatalf("unexpected error creating driver: %v", err)
+	}
+
+	ctx := context.Background()
+	n := 6
+	prefix := "/test-list-objects-v2"
+	var filePathes []string
```
Comment (milosgajdos) [inline registry/storage/driver/s3-aws/s3_test.go:383]: Can you please rename this variable to `filePaths`

**Review by milosgajdos — CHANGES_REQUESTED** [review]
Small variable name changes, otherwise this LGTM. I'll push to get this merged ASAP once we fix the var names.

File: registry/storage/driver/s3-aws/s3_test.go (hunk 6 above)
Comment (davidwtf) [reply registry/storage/driver/s3-aws/s3_test.go]: Renamed. 
PTAL

**Review by milosgajdos — APPROVED** [review]

**Comment by milosgajdos** [comment]
Hey @davidwtf can you please rebase?

**Comment by davidwtf** [comment]
> Hey @davidwtf can you please rebase?

done

**Comment by milosgajdos** [comment]
Closing in favour of https://github.com/distribution/distribution/pull/3624

**Pull request closed without merging**